run_tests.py: don't shard tests, use chunk size 1.

skvadrik opened the pull request:
Previously the script divided all tests into chunks of equal size, which
is suboptimal as different tests take different time, and often large
tests end up in the same chunk. This adds imbalance and makes the test
runner slower than it could be with chunk size 1.

Also, use a more pythonic way to gather results from threads. The old
bash script did that via per-thread log files because it is hard to do
it otherwise in bash. But in python one can simply return results from
each thread, avoiding the overhead on file I/O.

skvadrik commented:
@trofi Reducing chunk size to 1 as you suggested in https://github.com/skvadrik/re2c/pull/360#discussion_r654709590.

trofi commented on run_tests.py.in:
```diff
@@ -593,27 +560,25 @@ def main():
                       args.valgrind, args.wine,)
     ) as pool:
-        pool.map(run_pack, chunks(tests, int(len(tests) / jobs + 1)))
+        results = pool.map(run_test, tests)
```
As a tiny micro-optimization you can also switch to `imap` from `map`: https://stackoverflow.com/questions/26520781/multiprocessing-pool-whats-the-difference-between-map-async-and-imap.

As I understand it will allow for slightly faster output of first element. And will also at some later point turning `tests` itself into an iterable instead of an in-memory list.

skvadrik replied:
This small change took a few hours of debug. :D

First, it was necessary to change the way we get results: instead of `results = pool.map(...)` to have `for result in pool.imap(...)`, otherwise the script hangs on some futex and doesn't start running the tests:
```
$ time ./run_tests.py 
Running in 6 job(s)

```
strace looks like this:
```
strace -f -p $(ps a | grep run_tests | grep -v grep | egrep -o "^[0-9]+")
strace: Process 26953 attached
futex(0x558aaac2a270, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, FUTEX_BITSET_MATCH_ANY
```

Second, after I managed to get the tests running I started to have another problem: once in a few runs tet runner failed because an intermediate file got deleted from under a running thread. It turns out there was a race caused by two tests generating header with the same name. When I changed `map()` to `imap()` or even better `imap_unordered()` the two tests started to run in such a way that one of them removed the header while the other one was still processing it. One of the failed tests was a trivial modification of the other. It goes back to ancient times when it was useful, but not anymore, so I just removed the other test. Now I run test runner in a loop and it doesn't fail.

But it is a bad thing that intermediate files produced by different tests can collide in principle.

So, the end result is that run_tests.py use `imap_unordered`.

sergeyklay reviewed (changes requested):
`grep_count` is no longer needed here. Could you please remove it too?

skvadrik commented:
> `grep_count` is no longer needed here. Could you please remove it too?

Removed.

skvadrik commented:
@sergeyklay I rebased this on top of https://github.com/skvadrik/re2c/pull/363. The only change wrt. that PR is that I moved pattern compilation out of the function to the top level.

sergeyklay commented on run_tests.py.in:
```diff
@@ -301,12 +280,20 @@ def clean_test_tree(path):  # type: (str) -> None
 def filter_tests(path):  # type: (str) -> list
     """Filter and sort tests in ascending order."""
     test_files = [str(f) for f in Path(path).rglob('*.re')]
-    test_files.sort()
```
Bash version of test runner had this option. Don't we need it anymore?

skvadrik replied:
It doesn't make sense to sort them since we use `imap_unordered` anyway, does it?

sergeyklay commented on run_tests.py.in:
```diff
@@ -301,12 +280,20 @@ def clean_test_tree(path):  # type: (str) -> None
+type_header_pattern1 = re.compile(r'--type-header[ \t]*')
+type_header_pattern2 = re.compile(
+    r'.*--type-header[ \t]*([^ \t]*).*'
+)
```
Nitpic: Could you please use lowercase for variables in global scope? Also change the comment, we no longer have a cycle

skvadrik replied:
You mean uppercase, right?

sergeyklay replied:
Yes, I'm sorry, uppercase

skvadrik replied:
I uppercased them and fixed the comment.

sergeyklay commented on run_tests.py.in:
```diff
@@ -315,210 +302,183 @@ def run_pack(tests_pack):  # type: (list[str]) -> None
+    switches = switches.replace('$INPUT', outx)
+    switches = switches.replace('$OUTPUT', outy)
+    switches = switches.replace('$DEPFILE', f'{outy}.d')
+    switches = type_header_pattern1.sub(
```
I'm wonder if we can merge `type_header_pattern1 ` and `type_header_pattern2` and use the same pattern in bith cases. Did you try optimise this?

skvadrik replied:
I'd rather not do it in this PR, as it's unrelated.

sergeyklay reviewed (approved):
Looks solid and much better than previous approach

sergeyklay commented:
The `macos-clang-debug-intree-skeleton` job was stuck and failed after a hour or so. I rerun workflow
